### ISSUE TITLE
xt should be the first key in a magnet link

### ISSUE THREAD

**Issue by seaforium** [gausby/magnet#4]
I believe `xt` should come first so that the link starts with `magnet:?xt`. If starts with something else, say `magnet:?dn`, it is not being recognized by the torrent-client I tested it with (Deluge).

I wondered whether it is just the behavior of that particular program but then I found the following [here](http://bittorrent.org/beps/bep_0009.html) (under "magnet URI format"):
> The magnet URI format is:
> `v1: magnet:?xt=urn:btih:<info-hash>&dn=<name>&tr=<tracker-url>&x.pe=<peer-address>`
> `v2: magnet:?xt=urn:btmh:<tagged-info-hash>&dn=<name>&tr=<tracker-url>&x.pe=<peer-address>`

So it seems that the current way of handling it is not correct.

**Comment by gausby** [comment]
I cannot find any specification that specify the order of elements in a magnet link. The `xt` field is mandetory according to the [BEP0009](http://bittorrent.org/beps/bep_0009.html), *that is the only thing specified*.

For that reason I think it is okay to just ensure at least one `xt` is set—If they don't specify any order I don't think one should be enforced. So I think the package is fine as is.

**Comment by seaforium** [comment]
I think I was wrong. I have tested magnet links with different orders of keys with Transmission torrent client and it has actually correctly parsed them all no matter what the order of the keys was.

Now that I think about it, a magnet link is similar to an ordinary URL; and if one makes a GET request, the order of the key-value pairs in the query string doesn't matter. So why should it matter here? It looks like essentially the same thing. I don't know why I didn't think about this before and assumed there necessarily must be some order.

Sorry for taking your time. I really should have given it more thought and should have tried a different torrent client before opening up an issue.

**Comment by gausby** [comment]
Don't feel sorry—it is great to see someone who has taken interest in the project. Truth be told, I didn't really use this project. I did it while figuring out how to do a BitTorrent implementation in Elixir. I never completed it, though I was "pretty close."

Building a BitTorrent implementation is a great learning experience; Elixir is well suited for the task, though there are some difficulties here and there. Are you currently trying to build a BitTorrent implementation ?

**Comment by gausby** [comment]
Bencode is interesting in the sense that it shows how to do a recursive descend parser with pattern matching. Also, I am fairly sure that project is correct :)

Porting an old project to Elixir is a good task, but beware that Elixir is a different beast: It is powerful in a different way, and the tasks you can solve are different. I suggest that you try porting your project, and then I suggest that you try building some kind of network server—as this is where Elixir truly excels!

**Comment by seaforium** [comment]
Oh, I did a tutorial on exactly that topic — building a network server. It was a great experience, especially after I saw how OTP and Genserver help abstract all the details. But it is pretty hard for me to find things I would be passionate about that would involve OTP, message passing or parsing.

This is why my current plans are to port some utilities and then head to Phoenix. I hope to turn a useful CLI application into a web service that could be used by other people.
